### PR TITLE
Rework errors

### PR DESCRIPTION
**Pull request by jensdrenhaus**
Signed-off-by: Jens Drenhaus <jens.drenhaus@9elements.com>

### CODE REVIEW

File: trust/signature.go [hunk 1]
```diff
@@ -37,7 +39,7 @@ func (DummySigner) Sign(key crypto.PrivateKey, data []byte) ([]byte, error) {
 	sig := make([]byte, n)
 
 	if _, err := rand.Read(sig); err != nil {
-		return nil, fmt.Errorf("sign: %w", err)
+		return nil, fmt.Errorf("rand.DummySigner.Sign: %w: %s", ErrSigning, err)
```
Comment (MDr164) [inline trust/signature.go:42]: Not sure if `rand.DummySigner.Sign` is a good idea here. I assume `rand` is the package name (should be `trust` here then) but what if a user changes the import name of that package? Then `pkg.Struct.Method` might become confusing as the user imported it under a different name. I'd personally omit the pkg part and just name it `DummySigner.Sign` instead. What do you think?

File: trust/signature.go [hunk 2]
```diff
@@ -56,58 +58,35 @@ var _ Signer = RSAPSSSigner{}
 
 // Sign signes the provided data with the key named by privKey. The returned
 // byte slice contains a PSS signature value.
+// Problmes are reported by an error wrapping SigningError.
```
Comment (MDr164) [inline trust/signature.go:61]: Type: `Problmes`

File: trust/signature.go [hunk 3]
```diff
@@ -56,58 +58,35 @@ var _ Signer = RSAPSSSigner{}
...
-
-		return nil, ErrRSAPSSSigner
-	}
-
 	priv, ok := key.(*rsa.PrivateKey)
 	if !ok {
-		stlog.Debug("RSAPSSSigner: invalid key type %T", key)
-
-		return nil, ErrRSAPSSSigner
+		return nil, fmt.Errorf("trust.RSAPSSSigner.Sign: %w: %T, want rsa.PublicKey", ErrInvalidKey, key)
```
Comment (MDr164) [inline trust/signature.go:65]: See above regarding `pkg.Struct.Method`

File: trust/signature.go [hunk 4]
```diff
@@ -56,58 +58,35 @@ var _ Signer = RSAPSSSigner{}
...
 	}
 
 	opts := &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash}
 
 	ret, err := rsa.SignPSS(rand.Reader, priv, crypto.SHA256, data, opts)
 	if err != nil {
-		return nil, fmt.Errorf("sign: %w", err)
+		return nil, fmt.Errorf("trust.RSAPSSSigne.Sign: %w: %s", ErrSigning, err)
```
Comment (MDr164) [inline trust/signature.go:72]: see above

File: trust/signature.go [hunk 5]
```diff
@@ -56,58 +58,35 @@ var _ Signer = RSAPSSSigner{}
...
-
-		return ErrRSAPSSSigner
-	}
-
 	pub, ok := key.(*rsa.PublicKey)
 	if !ok {
-		stlog.Debug("RSAPSSSigner: invalid key type %T", key)
-
-		return ErrRSAPSSSigner
+		return fmt.Errorf("trust.RSAPSSSigner.Verify: %w: %T, want rsa.PublicKey", ErrInvalidKey, key)
```
Comment (MDr164) [inline trust/signature.go:82]: see above

File: trust/signature.go [hunk 6]
```diff
@@ -56,58 +58,35 @@ var _ Signer = RSAPSSSigner{}
...
 	}
 
 	opts := &rsa.PSSOptions{SaltLength: rsa.PSSSaltLengthEqualsHash}
 
 	err := rsa.VerifyPSS(pub, crypto.SHA256, hash, sig, opts)
 	if err != nil {
-		stlog.Debug("RSAPSSSigner: verification failed")
-
-		return ErrRSAPSSSigner
+		return fmt.Errorf("trust.RSAPSSSigner.Verify: %w", ErrVerification)
```
Comment (MDr164) [inline trust/signature.go:89]: see above

File: trust/signature.go [hunk 7]
```diff
@@ -118,49 +97,26 @@ type ED25519Signer struct{}
 var _ Signer = ED25519Signer{}
 
 // Sign signes the provided data with the key named by privKey.
+// Problmes are reported by an error wrapping SigningError.
```
Comment (MDr164) [inline trust/signature.go:100]: Same typo here

File: trust/signature.go [hunk 8]
```diff
@@ -118,49 +97,26 @@ type ED25519Signer struct{}
...
-
-		return nil, ErrED25519Signer
-	}
-
 	priv, ok := key.(ed25519.PrivateKey)
 	if !ok {
-		stlog.Debug("ED25519Signer: invalid key type %T", key)
-
-		return nil, ErrED25519Signer
+		return nil, fmt.Errorf("trust.ED25519Signer.Sign: %w: %T, want ed25519.PublicKey", ErrInvalidKey, key)
```
Comment (MDr164) [inline trust/signature.go:104]: see above

File: trust/signature.go [hunk 9]
```diff
@@ -118,49 +97,26 @@ type ED25519Signer struct{}
...
-
-		return ErrED25519Signer
-	}
-
 	pub, ok := key.(ed25519.PublicKey)
 	if !ok {
-		stlog.Debug("ED25519Signer: invalid key type %T", key)
-
-		return ErrED25519Signer
+		return fmt.Errorf("trust.ED25519Signer.Verify: %w: %T, want ed25519.PublicKey", ErrInvalidKey, key)
```
Comment (MDr164) [inline trust/signature.go:114]: see above

File: trust/signature.go [hunk 10]
```diff
@@ -118,49 +97,26 @@ type ED25519Signer struct{}
...
 	}
 
 	isValid := ed25519.Verify(pub, hash, sig)
 	if !isValid {
-		stlog.Debug("ED25519Signer: verification failed")
-
-		return ErrED25519Signer
+		return fmt.Errorf("trust.ED25519Signer.Verify: %w", ErrVerification)
```
Comment (MDr164) [inline trust/signature.go:119]: see above

File: stlog/syslog.go [hunk 11]
```diff
@@ -15,12 +16,14 @@ type kernelLogger struct {
 	level LogLevel
 }
 
+var errInitKlog = errors.New("init klog failed")
```
Comment (MDr164) [inline stlog/syslog.go:19]: There is no definite rule but usually errors like this should be exported. Especially for blackbox testing that would be mandatory but in our case it is optional. A linter should also not complain on exported package variable if it is an error.

Comment (jensdrenhaus) [reply stlog/syslog.go:19]: True. I guess this pkg needs further rework. E.g. putting the concrete logger implementations in own sub packages like stlog/klog or something. This klog thing is an implementation detail of the loggin package, and this particular error will not be populated to callers of stlog. This is why I left it unexported. Once it is in an own package, I would export it.

**Review by MDr164 — APPROVED** [review]

**Review by MDr164 — APPROVED** [review]

**Pull request merged**